### DISCUSSION TITLE
How does a Model have access to a class method like `where`?

### DISCUSSION THREAD

**Discussion by AlvesJorge** [mongodb/mongoid#5858] [Q&A] [answered]
As per the [docs](https://www.mongodb.com/docs/mongoid/current/tutorials/getting-started-rails7/), one should add `include Mongoid::Document` to their class to have any `new` document saved to the database, and overall have the ORM working as intended. 
So far so good, my app works, no problem there.

```ruby
class Band
 include Mongoid::Document
end
```

However my question is, according to [this](https://stackoverflow.com/questions/156362/what-is-the-difference-between-include-and-extend-in-ruby) and other stack overflow posts,  `include` makes the included class methods available to the target class instances, but it shouldn't do anything about adding the included class methods to the target class itself as class methods. 
So how can can we use something like `Band.where("...")` if `.where` isn't being inherited? 

Obviously it is being inherited I'm just trying to understand how, and where I went wrong with my logic :) Thanks

Context: My LSP can't pick up the `where` method on a class, so I have to go on a rabbit hole finding out why

**Comment by jamis** [comment, marked as answer]
Mongoid makes very heavy use of delegation and metaprogramming. It also uses `ActiveSupport::Concern` (from Rails) so that you can define class methods in a module that get applied when the module is included.

Looking specifically at the `#where` method will give you a bit of a taste of just how convoluted this gets. :)

* The `#where` method is defined on the `Mongoid::Criteria` class: [here](https://github.com/mongodb/mongoid/blob/master/lib/mongoid/criteria.rb#L420)
* The `#where` method is included in the set of methods that `Mongoid::Criteria::Queryable::Selectable` declares as "forwardable", [here](https://github.com/mongodb/mongoid/blob/master/lib/mongoid/criteria/queryable/selectable.rb#L945)
* `Mongoid::Document` includes `Mongoid::Composable` ([here](https://github.com/mongodb/mongoid/blob/master/lib/mongoid/document.rb#L27)), which extends `Mongoid::Findable` ([here](https://github.com/mongodb/mongoid/blob/master/lib/mongoid/composable.rb#L35)).
* `Mongoid::Findable` delegates those forwardable defined in `Mongoid::Criteria::Queryable::Selectable` to `with_default_scope` ([here](https://github.com/mongodb/mongoid/blob/master/lib/mongoid/findable.rb#L11-L14)).
* `Mongoid::Composable` (which you'll recall is included by `Mongoid::Document`) also includes `Mongoid::Scopable` ([here](https://github.com/mongodb/mongoid/blob/69eaf1d1edbd3998ebd0f062b445483b245bea3f/lib/mongoid/composable.rb#L55)), and `with_default_scope` is declared as a class method of `Mongoid::Scopable` ([here](https://github.com/mongodb/mongoid/blob/69eaf1d1edbd3998ebd0f062b445483b245bea3f/lib/mongoid/scopable.rb#L195)).
* `#with_default_scope` calls `queryable.with_default_scope`, and `#queryable` is also a class method of `Mongoid::Scopable` ([here](https://github.com/mongodb/mongoid/blob/69eaf1d1edbd3998ebd0f062b445483b245bea3f/lib/mongoid/scopable.rb#L107)).
* `#queryable` returns the active (in-scope) `Mongoid::Criteria` instance, or if no such instance is currently active, it instantiates a new `Mongoid::Criteria` instance ([here](https://github.com/mongodb/mongoid/blob/69eaf1d1edbd3998ebd0f062b445483b245bea3f/lib/mongoid/scopable.rb#L108)).
* Digging further, we see that `Mongoid::Criteria#with_default_scope` ([here](https://github.com/mongodb/mongoid/blob/69eaf1d1edbd3998ebd0f062b445483b245bea3f/lib/mongoid/criteria/scopable.rb#L152)) creates and returns a clone of the current criteria object, possibly initializing it with default scope conditions.

So, putting that all together, we can see that `Model.where` is delegated to the `#where` method of whatever `Mongoid::Criteria` instance is currently in scope (or a new instance if no such object exists). It's quite convoluted, and while it is one of the most complex paths in Mongoid right now (I believe), there are plenty of similar examples throughout the code base.

**Comment by AlvesJorge** [comment]
Thank you for such a detailed answer!
